tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -427,6 +434,27 @@ class ResolvedPair:
     resolved_to: NodeAndRange
 
 
+@dataclass
+class ScopedNodeAndRange:
+    node: chapel.AstNode
+    scopes: List[chapel.Scope] = field(default_factory=list)
+    rng: Range = field(init=False)
+
+    @staticmethod
+    def create(node: chapel.AstNode) -> Optional["ScopedNodeAndRange"]:
+        scopes = []
+        scope = node.scope()
+        while scope:
+            scopes.append(scope)
+            scope = scope.parent_scope()
+        if len(scopes) == 0:
+            return None
+        return ScopedNodeAndRange(node, scopes)
+
+    def __post_init__(self):
+        self.rng = location_to_range(self.node.location())
```

Comment:
Can't this be implicitly memoized using functools? I don't know the exact Python syntax, but:

```Python
@functools.memoize
def rng(self):
    location_to_range(self.node.location())
```

I think I prefer that to a "derived field" like `self.rng`.

Reply:
I tried `@functools.cache`, but it needs to hash the inputs and the `ScopedNodeAndRange` class is not easily hashable (at least without a custom hash function). For now I have `def rng` as an `@property` without the cache